Add kube-dns integration

aerostitch opened the pull request:
### What does this PR do?

Adds the kube-dns integration (grabs the metrics from prometheus using the new PrometheusCheck mother class recently merged.

### Motivation

I need to follow the dns metrics of my kubernetes platform, so instead of having to go to prometheus, I'd rather have everything in Datadog.

### Testing Guidelines

It's the standard basic techniques. The check and tests are really simple


@hkaj does this check fit what you had in mind for the usage of the PrometheusCheck mother class?

Thanks

hkaj commented:
Hey @aerostitch, thanks for contributing (again!)
Nice one, that's exactly what I had in mind :) I'll review it asap.

kmshultz commented:
Please add metrics metadata in metadata.csv - Thanks!

aerostitch commented:
Good catch @kmshultz.
Commit amended. PR updated.

Thanks! :)

kmshultz commented on kube_dns/README.md:
```diff
@@ -0,0 +1,35 @@
+
+## Configuration
+
+Edit the `prometheus.yaml` file to point to your server and port, set the masters to monitor
```
Shouldn't the file name be `kube_dns.yaml`?

aerostitch replied:
Oops, good catch indeed. Sorry for that.
Commit amended, PR updated.
Thanks! :)

kmshultz commented on kube_dns/metadata.csv:
```diff
@@ -0,0 +1,8 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+kubedns.response_size.bytes.sum,gauge,,response,byte,Size of the returns response in bytes.,0,kube_dns,response size sum
```
should be:

unit_name=byte
per_unit_name=<none>

aerostitch replied:
Fixing. Thanks! :)

kmshultz commented on kube_dns/metadata.csv:
```diff
@@ -0,0 +1,8 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+kubedns.response_size.bytes.sum,gauge,,response,byte,Size of the returns response in bytes.,0,kube_dns,response size sum
+kubedns.response_size.bytes.count,gauge,,response,,Number of requests on which the kubedns.response_size.bytes.sum metric is evaluated.,0,kube_dns,response size sample count
```
should unit_name be 'request', not 'response'?

aerostitch replied:
Technically it's requests for which a response was sent so it's response in the unit. Updating the description. Good catch! :)

kmshultz commented on kube_dns/metadata.csv:
```diff
@@ -0,0 +1,8 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+kubedns.response_size.bytes.sum,gauge,,response,byte,Size of the returns response in bytes.,0,kube_dns,response size sum
+kubedns.response_size.bytes.count,gauge,,response,,Number of requests on which the kubedns.response_size.bytes.sum metric is evaluated.,0,kube_dns,response size sample count
+kubedns.request_duration.seconds.sum,gauge,,request,second,Time (in seconds) each request took to resolve.,0,kube_dns,request duration time
```
should be:

unit_name=second, per_unit_name=(empty string)

aerostitch commented:
Thanks a lot for the review @kmshultz ! :)
Updated the metadata accordingly.

aerostitch commented:
I just noticed that the travis tests were failing due to mysql and totally unrelated failures.
Should it be re-ran as-is?

aerostitch commented:
Just pushed with a new note related to https://github.com/DataDog/dd-agent/pull/3344
Only the mysql check is failing with the error:
```
Failed task: RuntimeError Slave not up in time. Failing...
Skipping cleanup, disposable environments are great
rake aborted!
Slave not up in time. Failing...
/home/travis/build/DataDog/integrations-core/mysql/ci/mysql.rake:64:in `block (3 levels) in <top (required)>'
/home/travis/build/DataDog/integrations-core/mysql/ci/mysql.rake:101:in `block (4 levels) in <top (required)>'
/home/travis/build/DataDog/integrations-core/mysql/ci/mysql.rake:100:in `each'
/home/travis/build/DataDog/integrations-core/mysql/ci/mysql.rake:100:in `block (3 levels) in <top (required)>'
/home/travis/.rvm/gems/ruby-2.3.1/gems/datadog-sdk-testing-0.6.1/lib/tasks/sdk.rake:161:in `block (3 levels) in <top (required)>'
/home/travis/.rvm/gems/ruby-2.3.1/gems/datadog-sdk-testing-0.6.1/lib/tasks/sdk.rake:156:in `each'
/home/travis/.rvm/gems/ruby-2.3.1/gems/datadog-sdk-testing-0.6.1/lib/tasks/sdk.rake:156:in `block (2 levels) in <top (required)>'
/home/travis/.rvm/gems/ruby-2.3.1/gems/rake-11.3.0/exe/rake:27:in `<top (required)>'
/home/travis/.rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `eval'
/home/travis/.rvm/gems/ruby-2.3.1/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => ci:mysql:before_script
(See full trace by running task with --trace)
The command "bundle exec rake ci:run" exited with 1.
0.57s$ bundle exec rake requirements
/home/travis/.rvm/gems/ruby-2.3.1/gems/datadog-sdk-testing-0.6.1/lib/tasks/ci/hooks/pre-commit.py
No Github API set (missing creds?) cant crawl remotes.
No requirement version conflicts found. Looking good... ;)
```
Has that been fixed somewhere else?

irabinovitch commented:
@aerostitch Agreed this is unrelated to your changes. We'll dig in on our end.

hkaj commented on kube_dns/metadata.csv:
```diff
@@ -0,0 +1,8 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+kubedns.response_size.bytes.sum,gauge,,byte,,Size of the returns response in bytes.,0,kube_dns,response size sum
+kubedns.response_size.bytes.count,gauge,,response,,Number of responses on which the kubedns.response_size.bytes.sum metric is evaluated.,0,kube_dns,response size sample count
+kubedns.request_duration.seconds.sum,gauge,,request,second,Time (in seconds) each request took to resolve.,0,kube_dns,request duration time
```
Orientation should be `-1`, not `0`. Orientation is about what value we think is good for the metric and it goes like this:
```
- lower value is better --> -1
- higher value is better --> 1
- there's no better value --> 0
```

hkaj commented on kube_dns/metadata.csv:
```diff
@@ -0,0 +1,8 @@
+kubedns.response_size.bytes.sum,gauge,,byte,,Size of the returns response in bytes.,0,kube_dns,response size sum
+kubedns.response_size.bytes.count,gauge,,response,,Number of responses on which the kubedns.response_size.bytes.sum metric is evaluated.,0,kube_dns,response size sample count
+kubedns.request_duration.seconds.sum,gauge,,request,second,Time (in seconds) each request took to resolve.,0,kube_dns,request duration time
+kubedns.request_duration.seconds.count,gauge,,second,,Number of requests on which the kubedns.request_duration.seconds.sum metric is evaluated.,0,kube_dns,request duration sample
```
should the unit be request and not second then? (according to the description).

aerostitch replied:
Yeah I inverted the 2 lines. Good catch! :)

hkaj commented on kube_dns/metadata.csv:
```diff
@@ -0,0 +1,8 @@
+kubedns.request_duration.seconds.sum,gauge,,request,second,Time (in seconds) each request took to resolve.,0,kube_dns,request duration time
+kubedns.request_duration.seconds.count,gauge,,second,,Number of requests on which the kubedns.request_duration.seconds.sum metric is evaluated.,0,kube_dns,request duration sample
+kubedns.request_count.total,gauge,,request,,Number of DNS requests made.,0,kube_dns,total requests
+kubedns.error_count.total,gauge,,error,,Number of DNS requests resulting in an error.,0,kube_dns,requests errors
```
orientation should be -1

hkaj commented on kube_dns/metadata.csv:
```diff
@@ -0,0 +1,8 @@
+kubedns.request_duration.seconds.count,gauge,,second,,Number of requests on which the kubedns.request_duration.seconds.sum metric is evaluated.,0,kube_dns,request duration sample
+kubedns.request_count.total,gauge,,request,,Number of DNS requests made.,0,kube_dns,total requests
+kubedns.error_count.total,gauge,,error,,Number of DNS requests resulting in an error.,0,kube_dns,requests errors
+kubedns.cachemiss_count.total,gauge,,request,,Number of DNS requests that result in a cache miss.,0,kube_dns,cache miss
```
ditto

aerostitch commented on kube_dns/check.py:
```diff
@@ -0,0 +1,42 @@
+            'skydns_skydns_dns_request_count_total': 'request_count.total',
+            'skydns_skydns_dns_error_count_total': 'error_count.total',
+            'skydns_skydns_dns_cachemiss_count_total': 'cachemiss_count.total',
+        }
```
Makes sense. :)

aerostitch commented:
Thanks a lot for the review @hkaj !
Updated the PR accordingly.
Note that I also added an example on how to use it with service discovery in the `README.md` as it's probably how it's going to be used the most.

aerostitch commented:
The build failed again due to an unrelated mysql issue. @irabinovitch could you do your magic again so that it gets back to green please? Pretty please?! :)

Thanks

masci commented:
Here you go 🍏  apologies for the flakiness

hkaj commented on kube_dns/README.md:
```diff
@@ -0,0 +1,60 @@
+
+Edit the `kube_dns.yaml` file to point to your server and port, set the masters to monitor
+
+### Using with service discovery
```
Oh that's a good point. What about adding an auto_conf.yaml as well? [Here](https://github.com/DataDog/integrations-core/blob/5.13.2/redisdb/auto_conf.yaml) is an example for redis. We can use `k8s-dns-sidecar-amd64` as the common denominator in `docker_images`.

hkaj replied:
we'd also need to add `kubedns-amd64` in `docker_images` since CoreOS distributes it under that name.

aerostitch replied:
Hm... getting fancy here! ;)
Will do. Thanks

aerostitch replied:
added the auto config.
I also put `%%port_2%%` in there but the problem is that it might not always be port_2 if you want to disable the tcp listening for the dns for example.
Config example on the pod:
```                 "image": "gcr.io/google_containers/kubedns-amd64:1.9",                                                                                                                                             
                "imagePullPolicy": "IfNotPresent",                                                                                                                                                                 
                "livenessProbe": {                                                                                                                                                                                 
                    "failureThreshold": 5,                                                                                                                                                                         
                    "httpGet": {                                                                                                                                                                                   
                        "path": "/healthz-kubedns",                                                                                                                                                                
                        "port": 8080,                                                                                                                                                                              
                        "scheme": "HTTP"                                                                                                                                                                           
                    },                                                                                                                                                                                             
                    "initialDelaySeconds": 60,                                                                                                                                                                     
                    "periodSeconds": 10,                                                                                                                                                                           
                    "successThreshold": 1,                                                                                                                                                                         
                    "timeoutSeconds": 5                                                                                                                                                                            
                },                                                                                                                                                                                                 
                "name": "kubedns",                                                                                                                                                                                 
                "ports": [                                                                                                                                                                                         
                    {                                                                                                                                                                                              
                        "containerPort": 10053,                                                                                                                                                                    
                        "name": "dns-local",                                                                                                                                                                       
                        "protocol": "UDP"                                                                                                                                                                          
                    },                                                                                                                                                                                             
                    {                                                                                                                                                                                              
                        "containerPort": 10053,                                                                                                                                                                    
                        "name": "dns-tcp-local",                                                                                                                                                                   
                        "protocol": "TCP"                                                                                                                                                                          
                    },                                                                                                                                                                                             
                    {                                                                                                                                                                                              
                        "containerPort": 10055,                                                                                                                                                                    
                        "name": "metrics",                                                                                                                                                                         
                        "protocol": "TCP"                                                                                                                                                                          
                    }                                                                                                                                                                                              
                ],  ```

hkaj replied:
I think you can hardcode 10055 instead of port_2. It should be the most common port anyway and if a custom one is used, one can always override with annotation.

aerostitch replied:
Perfect, thanks @hkaj ! :)
Done

aerostitch commented:
Sounds good. Thanks! :)

deepakprabhakara commented:
Hello. Is there an official way to turn off kube dns monitoring? dd-agent is sending kube dns stats even though kubernetes stats itself is turned off.

aerostitch commented:
Not that I know of. The kubernetes, kubernetes_state and the kube_dns are 3 totally separate integrations.
Do you have a `kube_dns.yaml` file in your agent configuration @deepakprabhakara ? If you do, try to remove it. You didn't happen to put some annotations that get caught by the autodiscovery right?
https://docs.datadoghq.com/integrations/kube_dns/

deepakprabhakara commented:
@aerostitch Aah thanks, it must be autodiscovery.
